manuals/vslvmmount.1
.Dd July  5, 2026
.Dt VSLVMMOUNT 1
.Os
.Sh NAME
.Nm vslvmmount
.Nd mounts a Linux Logical Volume Manager (LVM) volume system
.Sh SYNOPSIS
.Nm vslvmmount
.Op Fl o Ar offset
.Op Fl X Ar extended_options
.Op Fl hvV
.Ar source
.Ar mount_point
.Sh DESCRIPTION
.Nm vslvmmount
is a utility to mount a Linux Logical Volume Manager (LVM) volume system
.Pp
.Nm vslvmmount
is part of the
.Nm libvslvm
package.
.Nm libvslvm
is a library to access the Linux Logical Volume Manager (LVM) volume system \
format
.Pp
.Ar source
a Linux Logical Volume Manager (LVM) volume system
.Ar mount_point
the directory to serve as mount point
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl h
shows this help
.It Fl o Ar offset
specify the volume offset in bytes
.It Fl v
verbose output to stderr, while vslvmmount will remain running in the foreground
.It Fl V
print version
.It Fl X Ar extended_options
extended options to pass to sub system
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# vslvmmount image1.raw /mnt/vslvm
vslvmmount 20260705
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr vslvminfo 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libvslvm/issues
.Sh COPYRIGHT
Copyright (C) 2014-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* check signature
  - check the first 4 sectors
  - update documentation regarding PV signature
* pyvslvm
  - complete segment object methods
  - complete stripe object methods
  - complete volume group object methods
* finish:
  - libvslvm_internal_logical_volume_read_buffer_from_file_io_handle
  - make sure to set the right offset in internal_logical_volume->chunks_vector 

Library:
* add support to specify metadata in external configuration file
* implement multi physical volume files support
  - map pv names to file IO pool?
* logical volume
  - expose creation_host
  - expose creation_time
  - implement multi stripe support
* metadata
  - parse and expose global values
  - make sure metadata is cleaned on close - also if no data files were set!
  - metadata: handle trailing comment in more places?
* API
  - allow for multiple input images (physical volumes)
  - allow open to read a text file with the metadata

Tools:
* vslvminfo
  - print stripes and in-volume offsets
* vslvmmount
  - allow use of logical volume names instead of lvm#?

Tests:
* add tests for library
* add tests for Python bindings
* add tests for tools

Deployment:
* update manuals

20141001
* see `git log' for more recent change log
* removed README.macosx

20140725
* worked on API

20140724
* worked on API

20140723
* worked on API

20140722
* worked on API
* moved metadata to separate code file

20140721
* worked on API

20140720
* worked on format support

20140201
* initial version based on libvsmbr



vslvmtools/vslvmmount.c
/*
 * Mounts a Linux Logical Volume Manager (LVM) volume system.
 *
 * Copyright (C) 2014-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <memory.h>
#include <system_string.h>
#include <types.h>

#include <stdio.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "mount_dokan.h"
#include "mount_fuse.h"
#include "mount_handle.h"
#include "vslvmtools_getopt.h"
#include "vslvmtools_i18n.h"
#include "vslvmtools_libcerror.h"
#include "vslvmtools_libclocale.h"
#include "vslvmtools_libcnotify.h"
#include "vslvmtools_libvslvm.h"
#include "vslvmtools_output.h"
#include "vslvmtools_signal.h"
#include "vslvmtools_unused.h"

mount_handle_t *vslvmmount_mount_handle = NULL;
int vslvmmount_abort                    = 0;

/* Signal handler for vslvmmount
 */
void vslvmmount_signal_handler(
      vslvmtools_signal_t signal VSLVMTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "vslvmmount_signal_handler";

	VSLVMTOOLS_UNREFERENCED_PARAMETER( signal )

	vslvmmount_abort = 1;

	if( vslvmmount_mount_handle != NULL )
	{
		if( mount_handle_signal_abort(
		     vslvmmount_mount_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal mount handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use vslvmmount to mount a Linux Logical Volume Manager (LVM) volume system.";

	vslvmtools_option_t options[ ] = {
		{ 'h', NULL, "shows this help" },
		{ 'o', "offset", "specify the volume offset in bytes" },
		{ 'v', NULL, "verbose output to stderr, while vslvmmount will remain running in the foreground" },
		{ 'V', NULL, "print version" },
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
		{ 'X', "extended_options", "extended options to pass to sub system" },
#endif
		{ 0, "volume", "a Linux Logical Volume Manager (LVM) volume system" },
		{ 0, "mount_point", "the directory to serve as mount point" },
	};
	system_character_t options_string[ 32 ];

	const system_character_t *path_prefix       = NULL;
	libvslvm_error_t *error                     = NULL;
	size_t path_prefix_size                     = 0;
	system_character_t *option_offset           = NULL;
	system_character_t *source                  = NULL;
	char *program                               = "vslvmmount";
	system_integer_t option                     = 0;
	int number_of_options                       = (int) ( sizeof( options ) / sizeof( vslvmtools_option_t ) );
	int verbose                                 = 0;

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) || defined( HAVE_LIBDOKAN )
	system_character_t *mount_point             = NULL;
	int result                                  = 0;
#endif

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	struct fuse_operations vslvmmount_fuse_operations;

	system_character_t *option_extended_options = NULL;

#if defined( HAVE_LIBFUSE3 )
	/* Need to set this to 1 even if there no arguments, otherwise this causes
	 * fuse: empty argv passed to fuse_session_new()
	 */
	char *fuse_argv[ 2 ]                        = { program, NULL };
	struct fuse_args vslvmmount_fuse_arguments  = FUSE_ARGS_INIT(1, fuse_argv);
#else
	struct fuse_args vslvmmount_fuse_arguments  = FUSE_ARGS_INIT(0, NULL);
	struct fuse_chan *vslvmmount_fuse_channel   = NULL;
#endif
	struct fuse *vslvmmount_fuse_handle         = NULL;

#elif defined( HAVE_LIBDOKAN )
	DOKAN_OPERATIONS vslvmmount_dokan_operations;
	DOKAN_OPTIONS vslvmmount_dokan_options;
#endif

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "vslvmtools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( vslvmtools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	vslvmtools_output_version_fprint(
	 stdout,
	 program );

	if( vslvmtools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = vslvmtools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				vslvmtools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'h':
				vslvmtools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'o':
				option_offset = optarg;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				vslvmtools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
			case (system_integer_t) 'X':
				option_extended_options = optarg;

				break;
#endif
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source volume.\n" );

		vslvmtools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind++ ];

	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing mount point.\n" );

		vslvmtools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) || defined( HAVE_LIBDOKAN )
	mount_point = argv[ optind ];
#endif
	libcnotify_verbose_set(
	 verbose );
	libvslvm_notify_set_stream(
	 stderr,
	 NULL );
	libvslvm_notify_set_verbose(
	 verbose );

	if( mount_handle_initialize(
	     &vslvmmount_mount_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize mount handle.\n" );

		goto on_error;
	}
	if( option_offset != NULL )
	{
		if( mount_handle_set_offset(
		     vslvmmount_mount_handle,
		     option_offset,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set volume offset.\n" );

			goto on_error;
		}
	}
#if defined( WINAPI )
	path_prefix = _SYSTEM_STRING( "\\LVM" );
#else
	path_prefix = _SYSTEM_STRING( "/lvm" );
#endif
	path_prefix_size = 1 + system_string_length(
	                        path_prefix );

	if( mount_handle_set_path_prefix(
	     vslvmmount_mount_handle,
	     path_prefix,
	     path_prefix_size,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to set path prefix.\n" );

		goto on_error;
	}
	if( mount_handle_open(
	     vslvmmount_mount_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source volume\n" );

		goto on_error;
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	if( option_extended_options != NULL )
	{
#if defined( HAVE_LIBFUSE3 )
		// fuse_opt_add_arg: Assertion `!args->argv || args->allocated' failed.
		vslvmmount_fuse_arguments.argc = 0;
		vslvmmount_fuse_arguments.argv = NULL;
#endif
		/* This argument is required but ignored
		 */
		if( fuse_opt_add_arg(
		     &vslvmmount_fuse_arguments,
		     "" ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
		if( fuse_opt_add_arg(
		     &vslvmmount_fuse_arguments,
		     "-o" ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
		if( fuse_opt_add_arg(
		     &vslvmmount_fuse_arguments,
		     option_extended_options ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
	}
	if( memory_set(
	     &vslvmmount_fuse_operations,
	     0,
	     sizeof( struct fuse_operations ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear fuse operations.\n" );

		goto on_error;
	}
	vslvmmount_fuse_operations.open       = &mount_fuse_open;
	vslvmmount_fuse_operations.read       = &mount_fuse_read;
	vslvmmount_fuse_operations.release    = &mount_fuse_release;
	vslvmmount_fuse_operations.opendir    = &mount_fuse_opendir;
	vslvmmount_fuse_operations.readdir    = &mount_fuse_readdir;
	vslvmmount_fuse_operations.releasedir = &mount_fuse_releasedir;
	vslvmmount_fuse_operations.getattr    = &mount_fuse_getattr;
	vslvmmount_fuse_operations.destroy    = &mount_fuse_destroy;

#if defined( HAVE_LIBFUSE3 )
	vslvmmount_fuse_handle = fuse_new(
	                          &vslvmmount_fuse_arguments,
	                          &vslvmmount_fuse_operations,
	                          sizeof( struct fuse_operations ),
	                          vslvmmount_mount_handle );

	if( vslvmmount_fuse_handle == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse handle.\n" );

		goto on_error;
	}
	result = fuse_mount(
	          vslvmmount_fuse_handle,
	          mount_point );

	if( result != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to fuse mount file system.\n" );

		goto on_error;
	}
#else
	vslvmmount_fuse_channel = fuse_mount(
	                           mount_point,
	                           &vslvmmount_fuse_arguments );

	if( vslvmmount_fuse_channel == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse channel.\n" );

		goto on_error;
	}
	vslvmmount_fuse_handle = fuse_new(
	                          vslvmmount_fuse_channel,
	                          &vslvmmount_fuse_arguments,
	                          &vslvmmount_fuse_operations,
	                          sizeof( struct fuse_operations ),
	                          vslvmmount_mount_handle );

	if( vslvmmount_fuse_handle == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse handle.\n" );

		goto on_error;
	}
#endif /* defined( HAVE_LIBFUSE3 ) */

	if( verbose == 0 )
	{
		if( fuse_daemonize(
		     0 ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable to daemonize fuse.\n" );

			goto on_error;
		}
	}
	result = fuse_loop(
	          vslvmmount_fuse_handle );

	if( result != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to run fuse loop.\n" );

		goto on_error;
	}
	fuse_unmount(
	 vslvmmount_fuse_handle );

	fuse_destroy(
	 vslvmmount_fuse_handle );

	fuse_opt_free_args(
	 &vslvmmount_fuse_arguments );

	return( EXIT_SUCCESS );

#elif defined( HAVE_LIBDOKAN )
	if( memory_set(
	     &vslvmmount_dokan_operations,
	     0,
	     sizeof( DOKAN_OPERATIONS ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear dokan operations.\n" );

		goto on_error;
	}
	if( memory_set(
	     &vslvmmount_dokan_options,
	     0,
	     sizeof( DOKAN_OPTIONS ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear dokan options.\n" );

		goto on_error;
	}
	vslvmmount_dokan_options.Version    = DOKAN_VERSION;
	vslvmmount_dokan_options.MountPoint = mount_point;

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	vslvmmount_dokan_options.SingleThread = TRUE;
#else
	vslvmmount_dokan_options.ThreadCount  = 0;
#endif
	if( verbose != 0 )
	{
		vslvmmount_dokan_options.Options |= DOKAN_OPTION_STDERR;
#if defined( HAVE_DEBUG_OUTPUT )
		vslvmmount_dokan_options.Options |= DOKAN_OPTION_DEBUG;
#endif
	}
/* This will only affect the drive properties
	vslvmmount_dokan_options.Options |= DOKAN_OPTION_REMOVABLE;
*/

#if ( DOKAN_VERSION >= 600 ) && ( DOKAN_VERSION < 800 )
	vslvmmount_dokan_options.Options |= DOKAN_OPTION_KEEP_ALIVE;

	vslvmmount_dokan_operations.CreateFile           = &mount_dokan_CreateFile;
	vslvmmount_dokan_operations.OpenDirectory        = &mount_dokan_OpenDirectory;
	vslvmmount_dokan_operations.CreateDirectory      = NULL;
	vslvmmount_dokan_operations.Cleanup              = NULL;
	vslvmmount_dokan_operations.CloseFile            = &mount_dokan_CloseFile;
	vslvmmount_dokan_operations.ReadFile             = &mount_dokan_ReadFile;
	vslvmmount_dokan_operations.WriteFile            = NULL;
	vslvmmount_dokan_operations.FlushFileBuffers     = NULL;
	vslvmmount_dokan_operations.GetFileInformation   = &mount_dokan_GetFileInformation;
	vslvmmount_dokan_operations.FindFiles            = &mount_dokan_FindFiles;
	vslvmmount_dokan_operations.FindFilesWithPattern = NULL;
	vslvmmount_dokan_operations.SetFileAttributes    = NULL;
	vslvmmount_dokan_operations.SetFileTime          = NULL;
	vslvmmount_dokan_operations.DeleteFile           = NULL;
	vslvmmount_dokan_operations.DeleteDirectory      = NULL;
	vslvmmount_dokan_operations.MoveFile             = NULL;
	vslvmmount_dokan_operations.SetEndOfFile         = NULL;
	vslvmmount_dokan_operations.SetAllocationSize    = NULL;
	vslvmmount_dokan_operations.LockFile             = NULL;
	vslvmmount_dokan_operations.UnlockFile           = NULL;
	vslvmmount_dokan_operations.GetFileSecurity      = NULL;
	vslvmmount_dokan_operations.SetFileSecurity      = NULL;
	vslvmmount_dokan_operations.GetDiskFreeSpace     = NULL;
	vslvmmount_dokan_operations.GetVolumeInformation = &mount_dokan_GetVolumeInformation;
	vslvmmount_dokan_operations.Unmount              = &mount_dokan_Unmount;

#else
	vslvmmount_dokan_operations.ZwCreateFile         = &mount_dokan_ZwCreateFile;
	vslvmmount_dokan_operations.Cleanup              = NULL;
	vslvmmount_dokan_operations.CloseFile            = &mount_dokan_CloseFile;
	vslvmmount_dokan_operations.ReadFile             = &mount_dokan_ReadFile;
	vslvmmount_dokan_operations.WriteFile            = NULL;
	vslvmmount_dokan_operations.FlushFileBuffers     = NULL;
	vslvmmount_dokan_operations.GetFileInformation   = &mount_dokan_GetFileInformation;
	vslvmmount_dokan_operations.FindFiles            = &mount_dokan_FindFiles;
	vslvmmount_dokan_operations.FindFilesWithPattern = NULL;
	vslvmmount_dokan_operations.SetFileAttributes    = NULL;
	vslvmmount_dokan_operations.SetFileTime          = NULL;
	vslvmmount_dokan_operations.DeleteFile           = NULL;
	vslvmmount_dokan_operations.DeleteDirectory      = NULL;
	vslvmmount_dokan_operations.MoveFile             = NULL;
	vslvmmount_dokan_operations.SetEndOfFile         = NULL;
	vslvmmount_dokan_operations.SetAllocationSize    = NULL;
	vslvmmount_dokan_operations.LockFile             = NULL;
	vslvmmount_dokan_operations.UnlockFile           = NULL;
	vslvmmount_dokan_operations.GetFileSecurity      = NULL;
	vslvmmount_dokan_operations.SetFileSecurity      = NULL;
	vslvmmount_dokan_operations.GetDiskFreeSpace     = NULL;
	vslvmmount_dokan_operations.GetVolumeInformation = &mount_dokan_GetVolumeInformation;
	vslvmmount_dokan_operations.Unmounted            = NULL;
	vslvmmount_dokan_operations.FindStreams          = NULL;
	vslvmmount_dokan_operations.Mounted              = NULL;

#endif /* ( DOKAN_VERSION >= 600 ) && ( DOKAN_VERSION < 800 ) */

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	DokanInit();
#endif
	result = DokanMain(
	          &vslvmmount_dokan_options,
	          &vslvmmount_dokan_operations );

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	DokanShutdown();
#endif
	switch( result )
	{
		case DOKAN_SUCCESS:
			break;

		case DOKAN_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: generic error\n" );
			break;

		case DOKAN_DRIVE_LETTER_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: bad drive letter\n" );
			break;

		case DOKAN_DRIVER_INSTALL_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unable to load driver\n" );
			break;

		case DOKAN_START_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: driver error\n" );
			break;

		case DOKAN_MOUNT_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unable to assign drive letter\n" );
			break;

		case DOKAN_MOUNT_POINT_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: mount point error\n" );
			break;

		default:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unknown error: %d\n",
			 result );
			break;
	}
	return( EXIT_SUCCESS );

#else
	fprintf(
	 stderr,
	 "No sub system to mount Linux Logical Volume Manager (LVM) format.\n" );

	return( EXIT_FAILURE );

#endif /* defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) */

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	if( vslvmmount_fuse_handle != NULL )
	{
		fuse_destroy(
		 vslvmmount_fuse_handle );
	}
	fuse_opt_free_args(
	 &vslvmmount_fuse_arguments );
#endif
	if( vslvmmount_mount_handle != NULL )
	{
		mount_handle_free(
		 &vslvmmount_mount_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

